### PR TITLE
 Add support for temporal types to batch importer

### PR DESCRIPTION
**Pull request by henriknyman**
- Date
- Time
- DateTime
- LocalTime
- LocalDateTime
- Duration

The default time zone for `Time` and `DateTime` is derived from the
database setting `dbms.db.timezone`.

### CODE REVIEW

File: community/csv/src/main/java/org/neo4j/csv/reader/Extractors.java [hunk 1]
```diff
@@ -952,6 +1008,186 @@ public AnyValue value()
         }
     }
 
+    public static class DateExtractor extends AbstractSingleValueExtractor<AnyValue>
```
Comment (sherfert) [inline community/csv/src/main/java/org/neo4j/csv/reader/Extractors.java]: It looks to me as if PointExtractor and all new Temporal extractors could share the field `value` and the methods `clear()` and `value()`. Maybe create an abstract type `AbstractSingleAnyValueExtractor extends AbstractSingleValueExtractor<AnyValue>` ?

File: community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/input/csv/CsvInputBatchImportIT.java [hunk 2]
```diff
@@ -177,7 +186,14 @@ public int bufferSize()
             InputEntity node = new InputEntity();
             node.id( UUID.randomUUID().toString(), Group.GLOBAL );
             node.property( "name", "Node " + i );
-            node.property( "point", "\"   { x : -4.2, y : " + i + ", crs: WGS-84 } \"");
+            node.property( "point", "\"   { x : -4.2, y : " + i + ", crs: WGS-84 } \"" );
+            node.property( "date", LocalDate.of( 2018, i % 12 + 1, i % 28 + 1 ) );
+            node.property( "time", OffsetTime.of( 1, i % 60, 0, 0, ZoneOffset.ofHours( 9 ) ) );
+            node.property( "dateTime",
```
Comment (sherfert) [inline community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/input/csv/CsvInputBatchImportIT.java:202]: I think we should have a dateTime without explicit timezone, too?

File: community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/input/csv/CsvInputBatchImportIT.java [hunk 3]
```diff
@@ -482,10 +504,26 @@ private void buildUpExpectedData(
...
             {
                 final Object expectedValue = node.propertyValue( i );
-                Consumer verify = actualValue ->
+                Consumer verify;
+                if ( expectedValue instanceof TemporalAmount )
                 {
-                    assertEquals( expectedValue, actualValue );
-                };
+                    // Since there is no straightforward comparison for TemporalAmount we add it to a reference
+                    // point in time and compare the result
```
Comment (sherfert) [inline community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/input/csv/CsvInputBatchImportIT.java:523]: I don't understand this. DurationValues implement equals in a sensible way.

File: community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/input/csv/CsvInputTest.java [hunk 4]
```diff
@@ -482,6 +488,156 @@ public void shouldParsePointPropertyValues() throws Exception
...
+        }
+    }
+
+    @Test
+    public void shouldParseDateTimePropertyValues() throws Exception
+    {
+        // GIVEN
+        DataFactory data = data(
+                ":ID,name,time:DateTime\n" +
+                "0,Mattias,2018-02-27T13:37\n" +
```
Comment (sherfert) [inline community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/input/csv/CsvInputTest.java:550]: I think we need to have a case that _does_ specify an explicit timezone here, too.

File: community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/input/csv/CsvInputTest.java [hunk 5]
```diff
@@ -482,6 +488,156 @@ public void shouldParsePointPropertyValues() throws Exception
...
+    }
+
+    @Test
+    public void shouldParseTimePropertyValues() throws Exception
+    {
+        // GIVEN
+        DataFactory data = data(
+                ":ID,name,time:Time\n" +
+                "0,Mattias,13:37\n" +
+                "1,Johan,\"16:20:01\"\n" );
```
Comment (sherfert) [inline community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/input/csv/CsvInputTest.java]: I think we need to have a case that does specify an explicit offset here, too.

**Review by sherfert — CHANGES_REQUESTED** [review]

File: community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/input/csv/CsvInputBatchImportIT.java (hunk 3 above)
Comment (henriknyman) [reply community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/input/csv/CsvInputBatchImportIT.java:523]: But `DurationValue` is not a publicly exposed type. I am trying to use only what you would get out of the Core API for verification.

**Review by sherfert — APPROVED** [review]

**Pull request merged**